File: receiver.py
import socket
import struct
import time
from collections import defaultdict
import os

class ImageReceiver:
    def __init__(self, host='127.0.0.1', port=8081):
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.sock.bind((host, port))
        self.buffer_size = 65535  # Max UDP packet size
        self.header_size = 8  # Size of PacketHeader struct in C++
        
        # Create output directory if it doesn't exist
        self.output_dir = 'data/received_images'
        os.makedirs(self.output_dir, exist_ok=True)
        
        print(f'Listening on {host}:{port}...')

    def parse_header(self, data):
        """Parse the packet header containing total_chunks and chunk_index."""
        return struct.unpack('II', data[:self.header_size])  # Two uint32_t values

    def receive_image(self):
        """Receive and reconstruct a complete image from chunks."""
        chunks = defaultdict(dict)  # Dictionary to store chunks for each image
        current_image_chunks = None
        expected_chunks = None
        
        while True:
            try:
                data, addr = self.sock.recvfrom(self.buffer_size)
                
                # Parse header
                total_chunks, chunk_index = self.parse_header(data)
                chunk_data = data[self.header_size:]
                
                # If this is the first chunk of a new image
                if chunk_index == 0:
                    current_image_chunks = total_chunks
                    expected_chunks = set(range(total_chunks))
                    
                # Store the chunk
                chunks[total_chunks][chunk_index] = chunk_data
                
                # Check if we have all chunks for the current image
                if current_image_chunks and len(chunks[current_image_chunks]) == current_image_chunks:
                    # Reconstruct the image
                    image_data = bytearray()
                    for i in range(current_image_chunks):
                        image_data.extend(chunks[current_image_chunks][i])
                    
                    # Save the image
                    timestamp = int(time.time() * 1000)  # Use milliseconds for unique filename
                    filename = f'{self.output_dir}/image_{timestamp}.jpg'
                    with open(filename, 'wb') as f:
                        f.write(image_data)
                    
                    print(f'Saved complete image: {filename} ({len(image_data)} bytes)')
                    
                    # Clear the chunks for this image
                    del chunks[current_image_chunks]
                    current_image_chunks = None
                    
            except Exception as e:
                print(f'Error receiving data: {e}')
                continue

    def run(self):
        """Start the receiver."""
        try:
            print('Starting image receiver...')
            self.receive_image()
        except KeyboardInterrupt:
            print('\nShutting down receiver...')
        finally:
            self.sock.close()

if __name__ == '__main__':
    receiver = ImageReceiver()
    receiver.run()

File: receiver.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using UnityEngine

public class ImageReceiver
{
    private readonly UdpClient udpClient;
    private const int BufferSize = 65535; // Max UDP packet size
    private const int HeaderSize = 8;     // Size of the packet header (two uint32_t)
    private readonly string outputDir = "data/received_images";

    private Dictionary<int, Dictionary<int, byte[]>> chunks = new Dictionary<int, Dictionary<int, byte[]>>();
    private int? currentImageChunks = null;
    private HashSet<int> expectedChunks = null;

    public ImageReceiver(string host = "127.0.0.1", int port = 8081)
    {
        udpClient = new UdpClient(port);
        Console.WriteLine($"Listening on {host}:{port}...");

        // Ensure output directory exists
        if (!Directory.Exists(outputDir))
        {
            Directory.CreateDirectory(outputDir);
        }
    }

    public void Run()
    {
        try
        {
            Console.WriteLine("Starting image receiver...");
            ReceiveImage();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Receiver error: {ex.Message}");
        }
        finally
        {
            udpClient.Close();
            Console.WriteLine("Shutting down receiver...");
        }
    }

    private void ReceiveImage()
    {
        IPEndPoint remoteEndPoint = new IPEndPoint(IPAddress.Any, 0);

        while (true)
        {
            try
            {
                byte[] data = udpClient.Receive(ref remoteEndPoint);

                // Parse the header
                (int totalChunks, int chunkIndex) = ParseHeader(data);

                // Extract the chunk data
                byte[] chunkData = new byte[data.Length - HeaderSize];
                Array.Copy(data, HeaderSize, chunkData, 0, chunkData.Length);

                // If this is the first chunk of a new image
                if (chunkIndex == 0)
                {
                    currentImageChunks = totalChunks;
                    expectedChunks = new HashSet<int>();
                    for (int i = 0; i < totalChunks; i++)
                    {
                        expectedChunks.Add(i);
                    }
                }

                // Add the chunk to the appropriate collection
                if (!chunks.ContainsKey(totalChunks))
                {
                    chunks[totalChunks] = new Dictionary<int, byte[]>();
                }

                chunks[totalChunks][chunkIndex] = chunkData;

                // Check if all chunks are received
                if (currentImageChunks.HasValue &&
                    chunks[currentImageChunks.Value].Count == currentImageChunks.Value)
                {
                    SaveCompleteImage(chunks[currentImageChunks.Value]);
                    chunks.Remove(currentImageChunks.Value);
                    currentImageChunks = null;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error receiving data: {ex.Message}");
            }
        }
    }

    private (int, int) ParseHeader(byte[] data)
    {
        // Read two integers (uint32) from the beginning of the data
        int totalChunks = BitConverter.ToInt32(data, 0);
        int chunkIndex = BitConverter.ToInt32(data, 4);
        return (totalChunks, chunkIndex);
    }

    private void SaveCompleteImage(Dictionary<int, byte[]> imageChunks)
    {
        byte[] imageData = CombineChunks(imageChunks);

        // Create a unique filename based on the current timestamp
        string timestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString();
        string filename = Path.Combine(outputDir, $"image_{timestamp}.jpg");

        // Save the image data to a file
        File.WriteAllBytes(filename, imageData);
        Console.WriteLine($"Saved complete image: {filename} ({imageData.Length} bytes)");
    }

    private byte[] CombineChunks(Dictionary<int, byte[]> imageChunks)
    {
        List<byte> combined = new List<byte>();

        for (int i = 0; i < imageChunks.Count; i++)
        {
            combined.AddRange(imageChunks[i]);
        }

        return combined.ToArray();
    }
}
